### PR TITLE
[17283] Fix issues in Dynamic Network Interfaces

### PR DESCRIPTION
**Pull request by MiguelCompany**
## Description

This PR adds a regression test based in Ubuntu containers for one long standing issue with dynamic network interfaces that was preventing communication when a participant was created with no network interfaces available.

It fixes the issue by correctly updating the locators on the proxy data of the participant, and all its readers or writers created with the default locators.

@Mergifyio backport 3.0.x 2.14.x 2.10.x

## Contributor Checklist

- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- _N/A_: Any new configuration API has an equivalent XML API (with the corresponding XSD extension) 
- [x] Changes are backport compatible: they do **NOT** break ABI nor change library core behavior. 
- [x] Changes are API compatible. 
- _N/A_: New feature has been added to the `versions.md` file (if applicable).
- _N/A_: New feature has been documented/Current behavior is correctly described in the documentation. 
    
- [x] Applicable backports have been included in the description.

## Reviewer Checklist

- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- _N/A_ If this is a critical bug fix, backports to the critical-only supported branches have been requested.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp [hunk 1]
```diff
@@ -1698,6 +1700,123 @@ void PDP::add_builtin_security_attributes(
 
 #endif // HAVE_SECURITY
 
+void PDP::local_participant_attributes_update_nts(
```
Comment (JesusPoderoso) [inline src/cpp/rtps/builtin/discovery/participant/PDP.cpp:1703]: As long as the method only updates locators and userData values, and not the remaining participant attributes, this method name may generate confusion.
What do you think about renaming it to be more specific?

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp [hunk 2]
```diff
@@ -1698,6 +1700,123 @@ void PDP::add_builtin_security_attributes(
...
+    // Update user data
+    auto participant_data = getLocalParticipantProxyData();
+    participant_data->m_userData.data_vec(new_atts.userData);
+
+    bool announce_locators = !mp_RTPSParticipant->is_intraprocess_only();
+    if (!announce_locators)
+    {
+        // If we are intraprocess only, we do not need to update locators
+        return;
+    }
```
Comment (JesusPoderoso) [inline src/cpp/rtps/builtin/discovery/participant/PDP.cpp:1715]: Why don't we simplify this?
```suggestion
    if (mp_RTPSParticipant->is_intraprocess_only())
    {
        // If we are intraprocess only, we do not need to update locators
        return;
    }
```

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp [hunk 3]
```diff
@@ -1698,6 +1700,123 @@ void PDP::add_builtin_security_attributes(
...
+    const auto& old_default_multicast = old_atts.defaultMulticastLocatorList;
+    const auto& new_default_unicast = new_atts.defaultUnicastLocatorList;
+    const auto& new_default_multicast = new_atts.defaultMulticastLocatorList;
+
+    // Early return if there is no change in default unicast locators
+    if ((old_default_unicast == new_default_unicast) &&
+            (old_default_multicast == new_default_multicast))
+    {
+        return;
+    }
```
Comment (JesusPoderoso) [inline src/cpp/rtps/builtin/discovery/participant/PDP.cpp:1768]: Why remaining locators are not considered?

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp [hunk 4]
```diff
@@ -1698,6 +1700,123 @@ void PDP::add_builtin_security_attributes(
...
+            writer->getAttributes().multicastLocatorList = new_default_multicast;
+            writer->getAttributes().unicastLocatorList = new_default_unicast;
+
+            WriterProxyData* wdata = nullptr;
+            GUID_t participant_guid;
+            wdata = addWriterProxyData(writer->getGuid(), participant_guid,
+                            [](WriterProxyData* proxy, bool is_update, const ParticipantProxyData& participant)
+                            {
+                                static_cast<void>(is_update);
+                                assert(is_update);
```
Comment (JesusPoderoso) [inline src/cpp/rtps/builtin/discovery/participant/PDP.cpp:1786]: Why?

**Review by JesusPoderoso — CHANGES_REQUESTED** [review]

File: test/dds/communication/PublisherMain.cpp [hunk 5]
```diff
@@ -29,9 +29,10 @@ using namespace eprosima::fastdds::dds;
  * --seed <int>
  * --wait <int>
  * --samples <int>
+ * --interval <int>
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/dds/communication/PublisherMain.cpp:33]: I have put them in the order they are processed below

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp (hunk 3 above)
Comment (MiguelCompany) [reply src/cpp/rtps/builtin/discovery/participant/PDP.cpp:1768]: Default locators are the ones that are copied to DATA(w) and DATA(r), so they are the only ones relevant for this method.

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp (hunk 4 above)
Comment (MiguelCompany) [reply src/cpp/rtps/builtin/discovery/participant/PDP.cpp:1786]: We are traversing the list of already existing writers, so their WriterProxyData should already exist, and thus would be updating.

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp (hunk 2 above)
Comment (JesusPoderoso) [reply src/cpp/rtps/builtin/discovery/participant/PDP.cpp:1715]: I like that, yes

**Review by JesusPoderoso — APPROVED** [review]
LGTM with green CI

**Comment by MiguelCompany** [comment]
@Mergifyio backport 3.0.x 2.14.x

**Pull request merged**